File: StructuralPattern/Composite/Composite.cc
#include <iostream>
#include <string>
#include <vector>
#include <memory>
#include <algorithm>
using namespace std;
class ComponentPtr
{
protected:
    std::string m_strName;

public:
    ComponentPtr(std::string str)
    {
        m_strName = str;
    }
    virtual void add(ComponentPtr *p) = 0;
    virtual void remove(ComponentPtr *p) = 0;
    virtual void display() = 0;
};

class LeafPtr : public ComponentPtr
{
public:
    LeafPtr(std::string str) : ComponentPtr(str) {}
    void add(ComponentPtr *p)
    {
        std::cout << "Leaf cannot add" << std::endl;
    }
    void remove(ComponentPtr *p)
    {
        std::cout << "Leaf cannot remove" << std::endl;
    }
    void display()
    {
        std::cout << m_strName << std::endl;
    }
};

class CompositePtr : public ComponentPtr
{
private:
    // 这里使用智能指针不用自己释放new的内存
    std::vector<std::shared_ptr<ComponentPtr>> m_vec;

public:
    CompositePtr(std::string str) : ComponentPtr(str){};
    ~CompositePtr()
    {
        if (!m_vec.empty())
        {
            m_vec.clear();
        }
    }
    void add(ComponentPtr *p)
    {
        auto it = find_if(m_vec.begin(), m_vec.end(),
                          [p](const shared_ptr<ComponentPtr> &ptr) { return p == ptr.get(); });
        if (it == m_vec.end())
            m_vec.push_back(shared_ptr<ComponentPtr>(p));
    }
    void remove(ComponentPtr *p)
    {
        auto it = find_if(m_vec.begin(), m_vec.end(),
                          [p](const std::shared_ptr<ComponentPtr> &ptr) { return p == ptr.get(); });
        if (it == m_vec.end())
            return;
        m_vec.erase(it);
    }
    void display()
    {
        for (auto it = m_vec.cbegin(); it != m_vec.cend(); it++)
        {
            (*it)->display();
        }
    }
};

int main()
{
    using namespace std;
    // 组合模式
    CompositePtr *p = new CompositePtr("总部");
    p->add(new LeafPtr("总部财务部门"));
    p->add(new LeafPtr("总部人力资源部门"));
    CompositePtr *p1 = new CompositePtr("上海分部");
    p1->add(new LeafPtr("上海分部财务部门"));
    p1->add(new LeafPtr("上海分部人力资源部门"));
    p->add(p1);
    p->display();
}

File: StructuralPattern/Composite/Composite.go
package main

import "fmt"

type Component interface {
	Add(Component)
	Remove(Component)
	DisPlay()
}
type Leaf struct {
	Name string
}

func (s *Leaf) Add(Component) {
	fmt.Println("leaf can not add")
}
func (s *Leaf) Remove(Component) {
	fmt.Println("leaf can not remove")
}
func (s *Leaf) DisPlay() {
	fmt.Println(s.Name)
}

type Composite struct {
	Name     string
	ComSlice []Component
}

func (c *Composite) Add(co Component) {
	b := true
	for _, v := range c.ComSlice {
		if v == co {
			b = false
		}
	}
	if b {
		c.ComSlice = append(c.ComSlice, co)
	}
}
func (c *Composite) Remove(co Component) {
	for i, v := range c.ComSlice {
		if v == co {
			c.ComSlice = append(c.ComSlice[:i-1], c.ComSlice[i+1:]...)
		}
	}
}
func (c *Composite) DisPlay() {
	for _, v := range c.ComSlice {
		v.DisPlay()
	}
}
func main() {
	p := Composite{Name: "总部"}
	p.Add(&Leaf{Name: "总部财务部门"})
	p.Add(&Leaf{Name: "总部人力资源部门"})

	p1 := Composite{Name: "上海分部"}
	p1.Add(&Leaf{Name: "上海分部财务部门"})
	p1.Add(&Leaf{Name: "上海分部人力资源部门"})

	p.Add(&p1)
	p.DisPlay()
}
